File: 3.x.x/crates-utopia/messaging/src/adapter/sms/plivo.rs
//! PHP `Utopia\Messaging\Adapter\SMS\Plivo`.

use serde_json::json;

use super::{sms_response_from_status, status_2xx, unknown_error, TYPE};
use crate::adapter::{expect_sms, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::SMS;
use crate::php::basic_auth;
use crate::response::ResponseData;

/// PHP `Adapter\SMS\Plivo`.
#[derive(Debug)]
pub struct Plivo {
    base: AdapterBase,
    auth_id: String,
    auth_token: String,
    from: Option<String>,
}

impl Plivo {
    /// PHP `__construct($authId, $authToken, $from = null)`.
    #[must_use]
    pub fn new(
        auth_id: impl Into<String>,
        auth_token: impl Into<String>,
        from: Option<String>,
    ) -> Self {
        Self {
            base: AdapterBase::default(),
            auth_id: auth_id.into(),
            auth_token: auth_token.into(),
            from,
        }
    }

    fn process_sms(&self, message: &SMS) -> ResponseData {
        let src = self
            .from
            .clone()
            .or_else(|| message.get_from().map(str::to_owned))
            .unwrap_or_else(|| "Plivo".into());
        let result = self.request_default(
            "POST",
            &format!("https://api.plivo.com/v1/Account/{}/Message/", self.auth_id),
            &[
                "Content-Type: application/x-www-form-urlencoded".into(),
                format!(
                    "Authorization: Basic {}",
                    basic_auth(&self.auth_id, &self.auth_token)
                ),
            ],
            Some(json!({
                "text": message.get_content(),
                "src": src,
                "dst": message.get_to().join("<"),
            })),
        );
        sms_response_from_status(message, &result, status_2xx, unknown_error)
    }
}

impl Adapter for Plivo {
    fn get_name(&self) -> &'static str {
        "Plivo"
    }
    fn get_type(&self) -> &'static str {
        TYPE
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::SMS
    }
    fn get_max_messages_per_request(&self) -> usize {
        1000
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(self.process_sms(expect_sms(message)?)))
    }
}

File: packages/messaging/src/Adapter/SMS/Plivo.php
<?php

namespace Utopia\Messaging\Adapter\SMS;

use Utopia\Messaging\Adapter\SMS as SMSAdapter;
use Utopia\Messaging\Messages\SMS as SMSMessage;
use Utopia\Messaging\Response;

// Reference Material
// https://www.plivo.com/docs/sms/api/message#send-a-message
class Plivo extends SMSAdapter
{
    protected const NAME = 'Plivo';

    /**
     * @param  string  $authId Plivo Auth ID
     * @param  string  $authToken Plivo Auth Token
     */
    public function __construct(
        private readonly string $authId,
        private readonly string $authToken,
        private readonly ?string $from = null,
    ) {
        parent::__construct();
    }

    public function getName(): string
    {
        return static::NAME;
    }

    public function getMaxMessagesPerRequest(): int
    {
        return 1000;
    }

    /**
     * {@inheritdoc}
     *
     * @throws \Exception
     */
    protected function process(SMSMessage $message): array
    {
        $response = new Response($this->getType());

        $result = $this->request(
            method: 'POST',
            url: "https://api.plivo.com/v1/Account/{$this->authId}/Message/",
            headers: [
                'Content-Type: application/x-www-form-urlencoded',
                'Authorization: Basic ' . base64_encode("{$this->authId}:{$this->authToken}"),
            ],
            body: [
                'text' => $message->getContent(),
                'src' => $this->from ?? $message->getFrom() ?? 'Plivo',
                'dst' => implode('<', $message->getTo()),
            ],
        );

        if ($result['statusCode'] >= 200 && $result['statusCode'] < 300) {
            $response->setDeliveredTo(\count($message->getTo()));
            foreach ($message->getTo() as $to) {
                $response->addResult($to);
            }
        } else {
            foreach ($message->getTo() as $to) {
                $response->addResult($to, 'Unknown error.');
            }
        }

        return $response->toArray();
    }
}
